src/test/java/io/supertokens/test/session/api/VerifySessionAPITest2_22.java

```diff
@@ -0,0 +1,257 @@
+/*
+ *    Copyright (c) 2023, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ *    This software is licensed under the Apache License, Version 2.0 (the
+ *    "License") as published by the Apache Software Foundation.
+ *
+ *    You may not use this file except in compliance with the License. You may
+ *    obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *    License for the specific language governing permissions and limitations
+ *    under the License.
+ */
+
+package io.supertokens.test.session.api;
+
+import com.auth0.jwt.interfaces.Claim;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.google.gson.JsonNull;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import io.supertokens.ProcessState;
+import io.supertokens.config.Config;
+import io.supertokens.test.TestingProcessManager;
+import io.supertokens.test.Utils;
+import io.supertokens.test.httpRequest.HttpRequestForTesting;
+import io.supertokens.utils.SemVer;
+import org.junit.*;
+import org.junit.rules.TestRule;
+
+import java.time.Instant;
+
+import static junit.framework.TestCase.*;
+import static junit.framework.TestCase.assertNull;
+import static org.junit.Assert.assertNotNull;
+
+public class VerifySessionAPITest2_22 {
+    @Rule
+    public TestRule watchman = Utils.getOnFailure();
+
+    @AfterClass
+    public static void afterTesting() {
+        Utils.afterTesting();
+    }
+
+    @Before
+    public void beforeEach() {
+        Utils.reset();
+    }
+
```

Comment:
Add tests for:
- Using v3 token with latest CDI, AccessTokenInfo points to public tenant.
- what happens when you give v4 access token to older cdi?
- what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh -> v4 token (with public tid)
- what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh (with cdi 2.21) -> v3 token (without public tid) - either update existing tests, or create new one.
- If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with latest CDI), should override tId with public.
- If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with CDI 2.21), should not affect tId.

Reply:
Using v3 token with latest CDI, AccessTokenInfo points to public tenant. --- DONE

what happens when you give v4 access token to older cdi? --- Works normally (added test)

what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh -> v4 token (with public tid) --- Doesn't force a refresh, but updates the version on refresh

what happens if you give a v2 access token to getSession -> it should give try refresh token -> refresh (with cdi 2.21) -> v3 token (without public tid) - either update existing tests, or create new one. --- similar to v4, doesn't force refresh, but updates version when refreshed. have updated existing tests

If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with latest CDI), should override tId with public. --- verify session succeeds and also uses public tenant only and not use the tId value. on refresh, we return UNAUTHORISED stating user data contains a protected field

If you have tId as a custom claim in a v3 access token, getSession should succeed, the claim not change. Refresh session (with CDI 2.21), should not affect tId. --- works as expected